starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala

```diff
@@ -0,0 +1,889 @@
+  ): Unit = {
+
+    if (builder.dryRun) {
+      logger.warn("** Reporter is running in 'DRY RUN' mode **")
+    }
+    try {
+      val metricData: List[MetricDatum] = new ArrayList[MetricDatum](
+        gauges.size + counters.size + 10 * histograms.size + 10 * timers.size
+      )
+      gauges.forEach { case (k: String, v: Gauge[_]) => processGauge(k, v, metricData) }
+      counters.forEach { case (k: String, v: Counter) => processCounter(k, v, metricData) }
+      histograms.forEach {
+        case (k: String, v: Histogram) => {
+          processHistogram(k, v, metricData)
+          processCounter(k, v, metricData)
+        }
+      }
+
+      meters.forEach {
+        case (k: String, v: Meter) => {
+          processMeter(k, v, metricData)
+          processCounter(k, v, metricData)
+        }
+      }
+      timers.forEach {
+        case (k: String, v: Timer) => {
+          processTimer(k, v, metricData)
+          processMeter(k, v, metricData)
+          processCounter(k, v, metricData)
+        }
+      }
+
+      val metricDataPartitions: Collection[List[MetricDatum]] =
+        CollectionsUtils.partition(metricData, maximumDatumsPerRequest)
+      val cloudWatchFutures: List[Future[PutMetricDataResult]] =
+        new ArrayList[Future[PutMetricDataResult]](metricData.size)
+
+      metricDataPartitions.forEach(partition => {
+        val putMetricDataRequest: PutMetricDataRequest = new PutMetricDataRequest()
+          .withNamespace(namespace)
+          .withMetricData(partition)
+        if (builder.dryRun) {
+          if (logger.isDebugEnabled) {
+            logger.debug("Dry run - constructed PutMetricDataRequest: {}", putMetricDataRequest)
+          }
+        } else {
+          cloudWatchFutures.add(cloudWatchAsyncClient.putMetricDataAsync(putMetricDataRequest))
+        }
+      })
+
+      cloudWatchFutures.forEach(cloudWatchFuture => {
+        try cloudWatchFuture.get
+        catch {
+          case e: Exception =>
+            logger.error(
+              "Error reporting metrics to CloudWatch. The data in this CloudWatch API request " +
+                "may have been discarded, did not make it to CloudWatch.",
+              e
+            )
+
+        }
+      })
+
+      if (logger.isDebugEnabled) {
+        logger.debug(
+          s"Sent ${metricData.size} datums to CloudWatch. Namespace: ${namespace}, metric data ${metricData}."
+        )
+      }
+    } catch {
+      case e: RuntimeException =>
+        logger.error("Error marshalling CloudWatch metrics.", e)
+
+    }
+  }
+
+  override def stop(): Unit = {
+    try super.stop()
+    catch {
+      case e: Exception => logger.error("Error when stopping the reporter.", e)
+
+    } finally if (!builder.dryRun) {
+      try cloudWatchAsyncClient.shutdown()
+      catch {
+        case e: Exception =>
+          logger.error("Error shutting down AmazonCloudWatchAsync", e)
+
+      }
+    }
+  }
+
+  private def processGauge[_ <: Numeric[_]](
```

Comment:
Is this scala Numeric? If that's the case, you should not do type casting here with `value.asInstanceOf[Double]` but instead use `implicitly[T].toDouble(value)`

Reply:
That's a very good catch! Thanks for that.

Reply:
@realstraw I took another look at this and rewrote it using more idiomatic Scala. Some notes: 

1. The method was trying to get a value from the gauge, which could be `null`, and so was wrapping it in an `Optional`. I used a Scala `Option` to do this now.
2. A Gauge is parameterized by type, but not necessarily a numeric type. The method now checks whether the non-null value returned by `getValue` is a `java.lang.Number` and then will do the cast from a Java `Number` to a Scala `Double`. 
3. In the `null` or non-numeric case, nothing will be reported and a warning will be logged.